// rxdb-hooks/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useRxQuery = exports.useRxDB = exports.useRxData = exports.useRxCollection = exports.Provider = exports.Context = void 0;
var context_1 = require("./context");
Object.defineProperty(exports, "Context", { enumerable: true, get: function () { return __importDefault(context_1).default; } });
__exportStar(require("./context"), exports);
var Provider_1 = require("./Provider");
Object.defineProperty(exports, "Provider", { enumerable: true, get: function () { return __importDefault(Provider_1).default; } });
__exportStar(require("./Provider"), exports);
var useRxCollection_1 = require("./useRxCollection");
Object.defineProperty(exports, "useRxCollection", { enumerable: true, get: function () { return __importDefault(useRxCollection_1).default; } });
__exportStar(require("./useRxCollection"), exports);
var useRxData_1 = require("./useRxData");
Object.defineProperty(exports, "useRxData", { enumerable: true, get: function () { return __importDefault(useRxData_1).default; } });
__exportStar(require("./useRxData"), exports);
var useRxDB_1 = require("./useRxDB");
Object.defineProperty(exports, "useRxDB", { enumerable: true, get: function () { return __importDefault(useRxDB_1).default; } });
__exportStar(require("./useRxDB"), exports);
var useRxQuery_1 = require("./useRxQuery");
Object.defineProperty(exports, "useRxQuery", { enumerable: true, get: function () { return __importDefault(useRxQuery_1).default; } });
__exportStar(require("./useRxQuery"), exports);
__exportStar(require("./plugins"), exports);
